File: nodns-bot-archive/internal/nostr/subscriber.go
package nostr

import (
	"context"
	"fmt"
	"log/slog"
	"sync"
	"time"

	"github.com/nbd-wtf/go-nostr"
	"nodns-bot/internal/config"
	"nodns-bot/internal/store"
)

// Subscriber manages connections to Nostr relays and forwards kind 11111 events.
type Subscriber struct {
	relays       []string
	zone         string
	logger       *slog.Logger
	store        *store.Store
	reconnectMin time.Duration
	reconnectMax time.Duration

	ctx    context.Context
	cancel context.CancelFunc
	wg     sync.WaitGroup
}

// NewSubscriber creates a new Nostr subscriber.
func NewSubscriber(cfg config.NostrConfig, s *store.Store, logger *slog.Logger) *Subscriber {
	ctx, cancel := context.WithCancel(context.Background())
	return &Subscriber{
		relays:       cfg.Relays,
		zone:         cfg.Zone,
		logger:       logger.With("component", "nostr-subscriber"),
		store:        s,
		reconnectMin: cfg.ReconnectMin,
		reconnectMax: cfg.ReconnectMax,
		ctx:          ctx,
		cancel:       cancel,
	}
}

// Subscribe connects to all configured relays and returns a channel of verified events.
func (sub *Subscriber) Subscribe() (<-chan *nostr.Event, error) {
	events := make(chan *nostr.Event, 256)

	lastSeen, err := sub.store.GetLastSeen()
	if err != nil {
		sub.logger.Warn("failed to get last_seen, starting from 0", "error", err)
		lastSeen = 0
	}

	since := nostr.Timestamp(lastSeen)
	filter := nostr.Filter{
		Kinds: []int{KindDNSRecord},
		Since: &since,
	}

	sub.logger.Info("subscribing to relays",
		"relays", sub.relays,
		"since", lastSeen,
		"zone", sub.zone,
	)

	for _, relayURL := range sub.relays {
		sub.wg.Add(1)
		go sub.maintainConnection(relayURL, filter, events)
	}

	return events, nil
}

// Stop shuts down all relay connections.
func (sub *Subscriber) Stop() {
	sub.logger.Info("stopping subscriber")
	sub.cancel()
	sub.wg.Wait()
}

// maintainConnection keeps a persistent connection to a single relay with exponential backoff.
func (sub *Subscriber) maintainConnection(relayURL string, filter nostr.Filter, events chan<- *nostr.Event) {
	defer sub.wg.Done()

	backoff := sub.reconnectMin

	for {
		select {
		case <-sub.ctx.Done():
			return
		default:
		}

		err := sub.runSubscription(relayURL, filter, events)
		if err == nil {
			return
		}

		sub.logger.Warn("relay subscription ended, reconnecting",
			"relay", relayURL,
			"error", err,
			"backoff", backoff,
		)

		select {
		case <-sub.ctx.Done():
			return
		case <-time.After(backoff):
			if backoff < sub.reconnectMax {
				backoff *= 2
			}
		}
	}
}

// runSubscription connects to a relay, subscribes, and forwards events until error or cancellation.
func (sub *Subscriber) runSubscription(relayURL string, filter nostr.Filter, events chan<- *nostr.Event) error {
	relay, err := nostr.RelayConnect(sub.ctx, relayURL)
	if err != nil {
		return fmt.Errorf("relay connect %s: %w", relayURL, err)
	}
	defer relay.Close()

	sub.logger.Info("connected to relay", "relay", relayURL)

	subCtx, subCancel := context.WithCancel(sub.ctx)
	defer subCancel()

	subscription, err := relay.Subscribe(subCtx, []nostr.Filter{filter})
	if err != nil {
		return fmt.Errorf("subscribe %s: %w", relayURL, err)
	}

	for {
		select {
		case <-subCtx.Done():
			return subCtx.Err()

		case evt, ok := <-subscription.Events:
			if !ok {
				return fmt.Errorf("events channel closed for %s", relayURL)
			}
			sub.logger.Debug("received event",
				"relay", relayURL,
				"event_id", evt.ID,
				"pubkey", evt.PubKey,
			)

			// Verify signature
			ok, err := evt.CheckSignature()
			if err != nil {
				sub.logger.Warn("signature check error",
					"event_id", evt.ID,
					"error", err,
				)
				continue
			}
			if !ok {
				sub.logger.Warn("invalid signature",
					"event_id", evt.ID,
					"pubkey", evt.PubKey,
				)
				continue
			}

			select {
			case events <- evt:
			case <-subCtx.Done():
				return subCtx.Err()
			}

		case <-subscription.EndOfStoredEvents:
			sub.logger.Debug("EOSE received", "relay", relayURL)

		case reason, ok := <-subscription.ClosedReason:
			if !ok {
				return fmt.Errorf("closed reason channel closed for %s", relayURL)
			}
			return fmt.Errorf("subscription closed by %s: %s", relayURL, reason)
		}
	}
}

File: nodns-bot-rs/src/subscriber.rs
//! Nostr relay subscriber — connects to multiple relays and forwards kind 11111/31111 events.
//!
//! Ported 1:1 from `nodns-bot/internal/nostr/subscriber.go`.
//!
//! Uses nostr-sdk's built-in relay pool (one shared `Client`) instead of
//! managing one client per relay.
//!
//! ## Relay reconnection (Issue #69)
//!
//! nostr-sdk 0.44 handles relay reconnection automatically. `Client::connect()`
//! spawns a background maintenance task per relay that monitors the WebSocket
//! and re-establishes dropped connections without caller intervention. The
//! `RelayPoolNotification::Shutdown` variant below corresponds to a *pool-wide*
//! shutdown (e.g. explicit `disconnect()` or process teardown), not a single
//! transient relay drop — those are recovered internally by the SDK.
//! Per-relay status is observable via `client.relays()` (used by the health
//! endpoint). There is no need for manual reconnection with exponential
//! backoff; the SDK's background task is always running while the client is
//! connected. If finer-grained status tracking is desired in the future, the
//! `Monitor` API (nostr-sdk 0.44+) can be attached to the `Client` builder to
//! stream `StatusChanged` notifications.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};

use nostr_sdk::{Client, Event, Filter, Kind, RelayMessage, RelayPoolNotification, Timestamp};
use thiserror::Error;
use tokio::sync::mpsc;
use tokio::task::JoinHandle;

use crate::config::NostrConfig;
use crate::store::Store;
use crate::types::{KIND_DNS_RECORD, KIND_DNS_REPLACEABLE};

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

#[derive(Debug, Error)]
pub enum SubscriberError {}

// ---------------------------------------------------------------------------
// Subscriber
// ---------------------------------------------------------------------------

/// Manages persistent connections to Nostr relays and forwards verified
/// kind 11111/31111 (DNS record) events through an [`mpsc`] channel.
pub struct Subscriber {
    client: Client,
    relays: Vec<String>,
    store: Arc<Store>,
    handle: Mutex<Option<JoinHandle<()>>>,
    stopped: Arc<AtomicBool>,
}

impl Subscriber {
    /// Create a subscriber with a pre-existing Client (shared with `AppState` for health checks).
    pub fn with_client(client: Client, cfg: &NostrConfig, store: Arc<Store>) -> Self {
        Self {
            client,
            relays: cfg.relays.clone(),
            store,
            handle: Mutex::new(None),
            stopped: Arc::new(AtomicBool::new(false)),
        }
    }

    /// Connect to all configured relays and return a channel of verified events.
    pub fn subscribe(&self) -> Result<mpsc::Receiver<Event>, SubscriberError> {
        let (tx, rx) = mpsc::channel(256);

        let last_seen = match self.store.get_last_seen() {
            Ok(ts) => ts,
            Err(e) => {
                tracing::warn!(error = %e, "failed to get last_seen, starting from 0");
                0
            }
        };

        let mut filter = Filter::new().kinds(vec![
            Kind::Custom(KIND_DNS_RECORD as u16),
            Kind::Custom(KIND_DNS_REPLACEABLE as u16),
        ]);
        if last_seen > 0 {
            filter = filter.since(Timestamp::from_secs(last_seen as u64));
        }

        tracing::info!(
            relays = ?self.relays,
            since = last_seen,
            "subscribing to relays"
        );

        let client = self.client.clone();
        let relays = self.relays.clone();
        let stopped = self.stopped.clone();

        let handle = tokio::spawn(async move {
            // Add all relays to the shared client.
            for url in &relays {
                if let Err(e) = client.add_relay(url).await {
                    tracing::warn!(relay = %url, error = %e, "failed to add relay");
                }
            }

            client.connect().await;

            tracing::info!("connected to relays, subscribing");

            if let Err(e) = client.subscribe(filter, None).await {
                tracing::error!(error = %e, "subscription failed");
                return;
            }

            // Forward verified events through the channel.
            let _ = client
                .handle_notifications(|notification| {
                    let tx = tx.clone();
                    let stopped = stopped.clone();

                    async move {
                        if stopped.load(Ordering::Relaxed) {
                            return Ok(true);
                        }

                        match notification {
                            RelayPoolNotification::Event { event, .. } => {
                                tracing::debug!(
                                    event_id = %event.id,
                                    pubkey = %event.pubkey,
                                    "received event"
                                );

                                if let Err(e) = event.verify() {
                                    tracing::warn!(
                                        event_id = %event.id,
                                        error = %e,
                                        "signature verification failed, skipping event"
                                    );
                                    return Ok(false);
                                }

                                let event = (*event).clone();
                                if tx.send(event).await.is_err() {
                                    return Ok(true);
                                }
                            }

                            RelayPoolNotification::Message { message, .. } => match &message {
                                RelayMessage::EndOfStoredEvents(_) => {
                                    tracing::debug!("EOSE received");
                                }
                                RelayMessage::Closed { message, .. } => {
                                    tracing::warn!(reason = %message, "subscription closed by relay");
                                    return Ok(true);
                                }
                                _ => {}
                            },

                            RelayPoolNotification::Shutdown => {
                                return Ok(true);
                            }
                        }

                        Ok(false)
                    }
                })
                .await;
        });

        *self.handle.lock().unwrap_or_else(|e| {
            tracing::error!("Handle mutex poisoned, recovering: {}", e);
            e.into_inner()
        }) = Some(handle);

        Ok(rx)
    }

    /// Gracefully shut down all relay connections.
    pub fn stop(&self) {
        tracing::info!("stopping subscriber");
        self.stopped.store(true, Ordering::SeqCst);

        if let Some(handle) = self
            .handle
            .lock()
            .unwrap_or_else(|e| {
                tracing::error!("Handle mutex poisoned, recovering: {}", e);
                e.into_inner()
            })
            .take()
        {
            handle.abort();
        }

        let client = self.client.clone();
        tokio::spawn(async move {
            client.disconnect().await;
        });
    }
}
